### ISSUE TITLE
Consider using fake pseudo-elements for anonymous box styling instead of modify_style_for_foo functions

### ISSUE THREAD

**Issue by pcwalton** [servo/servo#8570]
In Gecko, the equivalent of our `modify_style_for_foo` is accomplished via rules in the UA stylesheet with magic pseudo-elements that target anonymous boxes. This would be a cleaner way to handle anonymous box styling than what we do currently.

**Comment by jdm** [comment]
This is kind of sparse on any kind of background information or useful reading.

**Comment by mskrzypkows** [comment]
Could you explain something more about the task? Where ```modify_style_for_foo``` is used?

**Comment by pcwalton** [comment]
It's used to apply styles to anonymous boxes as generated by the CSS spec. Styles for those are by default inherited from the parent, but sometimes they need to have different styles from the parent in various ways. Right now the flow constructor notices this and calls functions with names beginning with `modify_style_for_` in order to apply the style changes. This is kind of a hard-coded UA stylesheet. It would be cleaner if there were actually a set of CSS rules in the UA stylesheet to handle these cases.

**Comment by jdm** [comment]
Looks like `generate_anonymous_table_flows_if_necessary` is an example of this.

**Comment by jmr0** [comment]
Would this require support from `selectors` if we took a similar approach? I took a quick glance and noticed `PseudoElement` only comes in after/before flavors

**Comment by pcwalton** [comment]
Yes, `selectors` would need to add support for these new pseudo-pseudo-elements. But I'm not sure they should actually be added to all the enums that contain before/after, because these pseudo-elements would not function as generated content.

**Comment by askobara** [comment]
Could I take this task?
I only not understand how to append pseudo class to the node. I have not found methods something like `add_class` in the docs.
Also should I implement [modify_style_for_text](https://github.com/servo/servo/blob/1d7689cb3fc6c8c6b9cdb3696a7b6ec744e48a3e/components/style/properties.mako.rs#L6934) as two separated pseudo-class and adds each of them separately according of conditions?

**Comment by pcwalton** [comment]
I don't understand what "I only not understand how to append pseudo class to the node. I have not found methods something like add_class in the docs." means. Could you rephrase?

Regarding `modify_style_for_text`, I think the first part (`relative`) shouldn't go through `modify_style` at all. Instead `position_offsets::{top,bottom,left,right}::to_computed_style()` should check the `position` in the `values::computed::Context`, as well as the pseudo-pseudo (which should also be added to the `Context`) and resolve to `Auto` in that case.

**Comment by askobara** [comment]
I assumed that needs to do the following:
* Add pseudo-classes like "::-moz-*", that contain styles from `modify_style_for_*` functions into the `resources/user-agent.css` file
* Append new pseudo-classes to nodes instead of changing original styles

Is it correct?

**Comment by pcwalton** [comment]
You don't need to actually modify any nodes. Just change the various accessors in layout to check the node type (the same notion of "node type" as in the DOM) and return which pseudo-pseudo should be used.

**Comment by schuster** [comment]
I'd like to take this on, assuming no one else is working on it.

**Comment by pcwalton** [comment]
Go for it! Any work in this direction would be valuable :)

**Comment by schuster** [comment]
I started with the styling for anonymous table rows, but after some debugging, it seems even the original modify_style_for* functions on these don't affect anything. I added a bit of println debugging and changed the function to instead set the pseudo-element's display style to `none` in this branch: https://github.com/schuster/servo/tree/anonymous-tables-debug

If I run that version of Servo on http://dbaron.org/css/test/sec170201 , all elements still appear to be rendered, even though many of them shouldn't be if my understanding is correct (because of the `display: none`). Am I missing something, or is that part of the code not having the intended effect?

In the meantime, I'll move on to some of the other modify_style_for functions.

**Comment by stshine** [comment]
Confirm that generate_anonymous_table_flows_if_necessary() doesn't take effect. But even if it works it won't work properly because it does not take siblings into account.
On the other hand, servo deal with `display: none` by not generate flow at all, while you have already generated the anonymous flows, so they will get displayed whatever.

**Comment by jdm** [comment]
There was a PR for this in #12496 which needs to be rebased and run through testing again.

**Comment by gabisurita** [comment]
If no one is working on that I'll like to give it I try :)

**Comment by schuster** [comment]
@pcwalton said in #13434 he already did this in #13870, so I think this issue should actually be closed.

**Comment by jdm** [comment]
Ooh, ok. Sorry @gabisurita! We can look for something else for you to work on if you would like!
